Added types of mongoDb to dependencies. MongoSecondaryEntityManager i…

adrianmjim opened the pull request:
…mplemented(it needs fixes)

Commit:
Added types of mongoDb to dependencies. MongoSecondaryEntityManager implemented(it needs fixes)

notaphplover commented on src/persistence/secondary/mongo-secondary-entity-manager.ts:
```diff
@@ -0,0 +1,62 @@
+  model: Model<TEntity>;
+  private collection: Collection;
+
+  constructor(url: string, dbName: string, collectionName: string) {
```
The constructor requires a model. The "collectionName" param is redundant, model.collectionName should be used instead. Maybe url and dbName could be extracted in an interface MongoBuildArgs.

notaphplover commented on src/persistence/secondary/mongo-secondary-entity-manager.ts:
```diff
@@ -0,0 +1,62 @@
+export class MongoSecondaryEntityManager<TEntity extends Entity> implements SecondaryEntityManager<TEntity> {
+
+  model: Model<TEntity>;
+  private collection: Collection;
```
Private and protected attributes in JS must start with underscore (for the sake of readability). Since TS code is compiled in JS I use to follow this rule in the project

notaphplover commented on src/persistence/secondary/mongo-secondary-entity-manager.ts:
```diff
@@ -0,0 +1,62 @@
+
+export class MongoSecondaryEntityManager<TEntity extends Entity> implements SecondaryEntityManager<TEntity> {
+
+  model: Model<TEntity>;
```
Maybe this attribute should be protected.

notaphplover commented on src/persistence/secondary/mongo-secondary-entity-manager.ts:
```diff
@@ -0,0 +1,62 @@
+    const results: Array<Promise<any>> = [];
+    for (const entity of entities) {
+      const mongoId = this.model.id;
+      results.push(this.collection.findOneAndUpdate({ id: mongoId }, entity));
```
Maybe we could achieve the updates with just one query using the updateMany method

notaphplover commented:
Lets fix syntax and compiler errors.

Commit:
Fixed lint errors.

Added modifiers to variables.
Added underscore to protected methods.
Added getter to the model.

Fixed an error in order of returning a Promise of multiple entities instead of multiple promises of an entity.

Commit:
Added toArray() calls in order of getting results from the cursor.

Commit:
Applied prettier suggestions.

Commit:
Added MongoBuildArgs interface. Wrapped mUpdate operations in one transaction. Added MongoClient to private attributes

notaphplover commented on src/persistence/secondary/mongo-secondary-entity-manager.ts:
```diff
@@ -0,0 +1,77 @@
+      });
+    } finally {
+      await session.endSession();
+    }
```
This code seems to be correct, but we can do the same with a promise oriented approach:

```Typescript
  public mUpdate(entities: TEntity[]): Promise<any> {
    const session = this._client.startSession();
    return session.withTransaction(() =>
      Promise.all(entities.map((entity) =>
        this._collection.findOneAndUpdate({id: this.model.id}, entity))),
    ).finally(session.endSession);
  }
```

notaphplover reviewed (commented):
The code seems to be right, we must add test before approving the PR.

Commit:
Added MongoBuildArgs interface. Added MongoClient to private attributes. Added users test. Added test declaration

Commit:
Added users.json

Commit:
Fixed linter errors.

Commit:
Test implemented. Collection based on promises to avoid async

Commit:
Merge remote-tracking branch 'origin/feature/implement-secondary-entity-manger' into feature/implement-secondary-entity-manger

# Conflicts:
#	src/test/mongo-secondary-entity-manager-test.ts

Commit:
All test implemented. Avoid async process

Commit:
All test refactored. Added MongoHelper. Erased mongoseed

Commit:
Fixed Prettier

Commit:
All test working

notaphplover reviewed (commented):
@Adrianmjim I've been running the test multiple times. the test "Mongo secondary entity manager test must update users" sometimes is run successfully and sometimes not. Could you try ti run multiple times the test on your machine in order to reproduce this behaviour?

Commit:
Fixed mUpdate.

An mUpdate operation is now expressed as a bult write operation: https://mongodb.github.io/node-mongodb-native/3.3/reference/ecmascriptnext/crud/

Commit:
Added prettier suggestions.

notaphplover commented on src/persistence/secondary/mongo-secondary-entity-manager.ts:
```diff
@@ -0,0 +1,77 @@
+  }
+
+  public async delete(id: string | number): Promise<any> {
+    return (await this._collection).findOneAndDelete({ id: id });
```
findOneAndDelete has an extra cost returning the removed document. Since we don't need the document, would it be a good idea to use the  "deleteOne" operation instead?

adrianmjim replied:
Yeah, that´s a great idea. I will remove it

Commit:
Fixed debug config.

Debug config will also remove the mongo db.

Commit:
Improved expect.

Numeric check can be performed using a toBe method.

Commit:
DB connection promise is used instead of collection promise.

MongoDB tutorials and StackOverflow pages use to reuse the db connection promise instead of the collection promise.
This approach solves some strange errors when multiple operations were performed in a certain order.

Commit:
Moved getter after the constructor.

Commit:
Added prettier suggestions.

Commit:
Mongo test has been changed.

Mongo find in doesn't seem to ensure a certain order of the results.
The mUpdate test has been changed in order of check the content without taking care of the result's order.

Commit:
Code refactor.

adrianmjim committed:
Merge pull request #17 from notaphplover/feature/improve-mupdate

Fixed mUpdate.

Commit:
UserCollection refactored. Fixed deleteOne. Fixed tests

Commit:
Date added to UserTest. Added initEntityManager

notaphplover commented on docker-compose.test.yml:
```diff
@@ -19,10 +19,13 @@ services:
     networks:
       - ant
+    ports:
+      - "27017:27017"
```
This map is usefull :). In order to reduce differences between docker-compose versions, I suggest to replicate this mapping in the docker-compose.ci.yml and docker-compose.debug.yml files. Could we use environment variables? I suggest MONGO_PORT_INNER and MONGO_PORT_OUTER:

```yaml
ports:
      - "${MONGO_PORT_OUTER}:${MONGO_PORT_INNER}"
```

notaphplover reviewed (commented):
Great job! 😃 

Now I think it's time to refactor this in order to use the id field of the model's instance as the id field. Maybe the Model class should be refactored. We can also merge this and do the refactor in a new PR, what do you think @Adrianmjim?

notaphplover commented:
Fixes #3

Commit:
Refactored id field of mongo to match with Id model

Commit:
Added Mongo ports to .env

notaphplover reviewed (approved):
I think we are ready. We can merge this and refactor the AntModel in a new PR.
